external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -342,80 +415,103 @@ private[kafka010] object CachedKafkaConsumer extends Logging {
+        logWarning(s"Attempting to release consumer that is not in the cache")
       }
     }
   }
 
   /**
    * Get a cached consumer for groupId, assigned to topic and partition.
    * If matching consumer doesn't already exist, will be created using kafkaParams.
+   * The returned consumer must be released explicitly using [[KafkaDataConsumer.release()]].
+   *
+   * Note: This method guarantees that the consumer returned is not currently in use by any one
+   * else. Within this guarantee, this will make a best effort attempt to re-use consumers by
+   * caching them and tracking when they are in use.
    */
-  def getOrCreate(
-      topic: String,
-      partition: Int,
-      kafkaParams: ju.Map[String, Object]): CachedKafkaConsumer = synchronized {
-    val groupId = kafkaParams.get(ConsumerConfig.GROUP_ID_CONFIG).asInstanceOf[String]
-    val topicPartition = new TopicPartition(topic, partition)
-    val key = CacheKey(groupId, topicPartition)
-
-    // If this is reattempt at running the task, then invalidate cache and start with
-    // a new consumer
+  def acquire(
+      topicPartition: TopicPartition,
+      kafkaParams: ju.Map[String, Object],
+      useCache: Boolean): KafkaDataConsumer = synchronized {
+    val key = new CacheKey(topicPartition, kafkaParams)
+    val existingInternalConsumer = cache.get(key)
+
+    lazy val newInternalConsumer = new InternalKafkaConsumer(topicPartition, kafkaParams)
+
     if (TaskContext.get != null && TaskContext.get.attemptNumber >= 1) {
-      removeKafkaConsumer(topic, partition, kafkaParams)
-      val consumer = new CachedKafkaConsumer(topicPartition, kafkaParams)
-      consumer.inuse = true
-      cache.put(key, consumer)
-      consumer
-    } else {
-      if (!cache.containsKey(key)) {
-        cache.put(key, new CachedKafkaConsumer(topicPartition, kafkaParams))
+      // If this is reattempt at running the task, then invalidate cached consumer if any and
+      // start with a new one.
+      if (existingInternalConsumer != null) {
+        if (existingInternalConsumer.inuse) {
+          // Consumer exists in cache and is somehow in use. Don't close it immediately, but
+          // mark it for being closed when it is released.
+          existingInternalConsumer.markedForClose = true
+          NonCachedKafkaDataConsumer(newInternalConsumer)
+
+        } else {
+          // Consumer exists in cache and is not in use, so close it immediately and replace
+          // it with a new one.
+          existingInternalConsumer.close()
+          cache.put(key, newInternalConsumer)
+          CachedKafkaDataConsumer(newInternalConsumer)
+
+        }
+      } else {
+        // Consumer is not cached, put the new one in the cache
+        cache.put(key, newInternalConsumer)
+        CachedKafkaDataConsumer(newInternalConsumer)
+
       }
-      val consumer = cache.get(key)
-      consumer.inuse = true
-      consumer
+    } else if (!useCache) {
```

Comment:
this `if` should be moved before the above `if`.

Reply:
why? I am saying that we should NOT reuse consumers for ANY task retries, independent of the desire to use the cache or not.

Reply:
When `useCache` is `false`, i would expect `newInternalConsumer` should never be put into the cache. The above `if` may put `newInternalConsumer` into the cache. If we put a consumer used by a continuous processing query into the cache and assume it never ends, it will prevent other micro batch queries from putting a consumer reading the same topic partition into the cache.

Reply:
Technically that wont happens because the continuous query and the batch query will have different groupids. But I agree that if `useCache` is `false`, then we should not put it in the cache in any way. In fact, we can simplify the task retry case further by never putting the new one in the cache, only invalidate the existing cached one.